### PR TITLE
[CHORE] Project migrated to Swift 4.1

### LINKED ISSUE #1466: [CHORE] Migrate the project to Swift 4.1
Swift 4.1 has been released with the latest update of Xcode (9.3), bringing some new language features and deprecating others. This warrants the project to be migrated over to the newer version of Swift.

Here is a list of the notable features:
- Deprecated `flatMap` in favor of `compactMap` which provides the same functionality under a different and more descriptive name.
[SE-0187 Introduce Sequence.compactMap(_:)](https://github.com/apple/swift-evolution/blob/master/proposals/0187-introduce-filtermap.md)

- Synthesizing `Equatable` and `Hashable` conformance for structs and enums in a way that is similar to `Codable` ([Synthesis in Extensions](https://github.com/apple/swift-evolution/blob/master/proposals/0185-synthesize-equatable-hashable.md#synthesis-in-extensions))
[SE-0185 Synthesizing Equatable and Hashable conformance](https://github.com/apple/swift-evolution/blob/master/proposals/0185-synthesize-equatable-hashable.md)

For the complete list of features visit https://swift.org/blog/swift-4-1-released/

### PR DESCRIPTION
**Pull request by Sameesunkaria**
@RocketChat/ios

The project has been migrated to Swift 4.1. It goes without saying that Xcode 9.3 will be required to build this project. The changes made have been documented below.

- The migration mostly includes changing `flatMap` to `compactMap`.
- The `identifier_name` rule for SwiftLint has been disabled, because the rule now applies to local constants/variables. This causes an error for local variable names that start with an underscore (since, `allowed_symbols: "_"` will not work for `identifier_name` we cannot add an exception for underscore).<br>
 **This should be properly addressed in the future.** <br>
 refer to: realm/SwiftLint#1852 and realm/SwiftLint#1762

- The Circle CI config has been updated. Currently they do not support iOS 10.3 devices with an Xcode 9.3 build, so those devices have been commented out.

**Miscellaneous:**
- Fixed the date from string test (`testDateFromString`) in `DateExtension` to work in all time zones.

Note: This contains some commits from #1464. Not sure what I can do about that.

Closes #1466

### CODE REVIEW

**Comment by Sameesunkaria** [comment]
Not sure whats happening here, but sometimes the async tests are taking much longer than they should. (and hence failing)
This has happened to 
- `testFetchCommands` 3b7aa76
- `testCatastrophicBacktrackingOfRegex` 58996bb

The tests *should* pass with a rebuild on circle ci.
We could consider relaxing the time for expectations as well.

**Comment by rafaelks** [comment]
@Sameesunkaria I did fix all the names instead of silencing the SwiftLint for now, I think that's the best way to go... the PR is #1460. Let's wait for the merge, then we can merge your changes on the flatMap, OK?

File: .swiftlint.yml [hunk 1]
```diff
@@ -1,11 +1,9 @@
+disabled_rules:
+  - identifier_name
```
Comment (rafaelks) [inline .swiftlint.yml]: This is fixed already.

**Comment by Sameesunkaria** [comment]
@rafaelks Everything should be fine now.

**Pull request merged**